numpy/core/src/umath/ufunc_object.c

```diff
@@ -1748,10 +1763,148 @@ make_arr_prep_args(npy_intp nin, PyObject *args, PyObject *kwds)
+
+        core_ndim = ufunc->core_num_dims[i];
+        if (op[i] != NULL) {
+            op_ndim = PyArray_NDIM(op[i]);
+        }
+        else {
+            op_ndim = broadcast_ndim + core_ndim;
+        }
+        item = PyList_GetItem(axes, i);
+        /* if not a tuple already, make it one; content is checked later */
+        if (!PyTuple_Check(item)) {
+            item = PyTuple_Pack(1, item);
+            if (item == NULL) {
+                return -1;
+            }
+        }
+        else {
+            Py_INCREF(item);
+        }
+        if (PyTuple_Size(item) != core_ndim) {
+            if (core_ndim == 1) {
+                PyErr_Format(PyExc_TypeError,
+                             "axes item %d should be a tuple with a "
+                             "single element, or an integer", i);
+            }
+            else {
+                PyErr_Format(PyExc_TypeError,
+                             "axes item %d should be a tuple with %d "
+                             "elements", i, core_ndim);
+            }
+            goto fail;
+        }
+        /*
+         * Now create the remap, starting with the core dimensions, and
+         * then adding the remaining axes that are to be iterated over.
+         */
+        for (j = 0; j < core_ndim; j++) {
+            requested_axis = (int)PyNumber_AsSsize_t(
+                PyTuple_GetItem(item, j), NULL);
+            if (error_converting(requested_axis) ||
+                (check_and_adjust_axis(&requested_axis, op_ndim) < 0)) {
+                goto fail;
+            }
+            if (have_seen_axis[requested_axis]) {
+                PyErr_Format(PyExc_ValueError,
+                             "axes item %d has value %d repeated",
+                             i, requested_axis);
+                goto fail;
+            }
+            have_seen_axis[requested_axis] = 1;
+            remap_axis[i][op_ndim - core_ndim + j] = requested_axis;
+        }
+        Py_DECREF(item);
+        /*
+         * Fill out the ndim-core_ndim axes that have not yet been set,
+         * using have_seen_axis to skip over the core_ndim entries set above.
+         */
+        for (j = 0, iter_axis = 0; j < op_ndim - core_ndim; j++) {
+            while (have_seen_axis[iter_axis]) {
+                iter_axis++;
+            }
+            remap_axis[i][j] = iter_axis++;
+        }
+        /*
+         * Check whether we are actually remapping anything. Here,
+         * iter_axis can only equal j if all broadcast axes were the same.
+         */
+        if (iter_axis == j) {
+            while (j < op_ndim && remap_axis[i][j] == j) {
+                j++;
+            }
+        }
+        if (j == op_ndim) {
+            remap_axis[i] = NULL;
+        }
+    } /* end of for(i) loop over operands */
+    return 0;
+
+fail:
+    Py_DECREF(item);
+    return -1;
+}
+
+#define REMAP_AXIS(op, axis) ((remap_axis != NULL && remap_axis[op] != NULL)? \
```

Comment:
I suppose if the above change to `iop` or similar, this should too.

Reply:
done